src/Silk.Core.Discord/EventHandlers/Guilds/GuildEventHandlers.cs

```diff
@@ -0,0 +1,26 @@
+﻿using System.Threading;
+using System.Threading.Tasks;
+using MediatR;
+using Microsoft.Extensions.Logging;
+using Silk.Core.Discord.EventHandlers.Notifications;
+
+namespace Silk.Core.Discord.EventHandlers.Guilds
+{
+    public class GuildEventHandlers : INotificationHandler<GuildAvailable>, INotificationHandler<GuildCreated>, INotificationHandler<GuildDownloadCompleted>
+    {
+        private readonly GuildEventHandlerService _guildHandler;
+        private readonly ILogger<GuildEventHandlers> _logger;
+
+        public GuildEventHandlers(GuildEventHandlerService guildHandler, ILogger<GuildEventHandlers> logger)
+        {
+            _guildHandler = guildHandler;
+            _logger = logger;
+        }
+
+        public async Task Handle(GuildCreated notification, CancellationToken cancellationToken) =>
+            _guildHandler.CacheQueue.Enqueue(new(_guildHandler.JoinedGuild(notification)));
+        public async Task Handle(GuildAvailable notification, CancellationToken cancellationToken) =>
+            _guildHandler.CacheQueue.Enqueue(new(() => _guildHandler.CacheGuildAsync(notification.Args.Guild, notification.Client.ShardId)));
+        public async Task Handle(GuildDownloadCompleted notification, CancellationToken cancellationToken) => _guildHandler.MarkCompleted(notification.Client.ShardId);
+    }
```

Comment:
The entire point of Mediator is to provide tooling like this I'm pretty sure. It handles the queueing and the handling. It's basically like using RabbitMQ but in-process.

Reply:
MediatR offers pipelining, yes. I'm however both incompetent and lacking the time to invest into research for that, though.

Reply:
Though I wouldn't call this CQRS, which is what MediatR actually tackles.

Reply:
> MediatR is a low-ambition library trying to solve a simple problem — decoupling the in-process sending of messages from handling messages.

MediatR is designed to handle message queuing. As I said, it's the equivalent of RabbitMQ but in-process.

Reply:
I'm essentially doing that, but with a twist. The handler for a given message enqueues a lazy task in a service. It's not great, but, it works.